trio/monitor.py

```diff
@@ -0,0 +1,380 @@
+import inspect
+import os
+import random
+import signal
+import string
+import sys
+import traceback
+from types import FunctionType
+
+from async_generator._impl import ANextIter
+
+from . import Queue, WouldBlock, BrokenStreamError
+from ._highlevel_serve_listeners import _run_handler
+from ._version import __version__
+from .abc import Instrument
+from .hazmat import current_task, Task
+
+# inspiration: https://github.com/python-trio/trio/blob/master/notes-to-self/print-task-tree.py
+# additional credit: python-trio/trio#429 (You are being ratelimited)
+
+
+def walk_coro_stack(coro):
+    while coro is not None:
+        if hasattr(coro, "cr_frame"):
+            # A real coroutine
+            yield coro.cr_frame, coro.cr_frame.f_lineno
+            coro = coro.cr_await
+        elif isinstance(coro, ANextIter):
+            # black hole
+            return
+        else:
+            # A generator decorated with @types.coroutine
+            yield coro.gi_frame, coro.gi_frame.f_lineno
+            coro = coro.gi_yieldfrom
+
+
+class Monitor(Instrument):
```

Comment:
Thought: another way to structure this would be to move the main loop into its own class or top-level function, and have the instrument as a separate class. When you ran the `monitor` command, it would create a new instrument object like `instrument = MonitorTrace(myqueue)` or whatever, make it an active instrument (`trio.hazmat.add_instrument`), read its output until done, then deactivate it again (`trio.hazmat.remove_instrument`).